numpy/core/include/numpy/npy_common.h

```diff
@@ -361,49 +363,202 @@ typedef double npy_double;
 typedef Py_hash_t npy_hash_t;
 #define NPY_SIZEOF_HASH_T NPY_SIZEOF_INTP
 
-/*
- * Disabling C99 complex usage: a lot of C code in numpy/scipy rely on being
- * able to do .real/.imag. Will have to convert code first.
- */
-#if 0
-#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_DOUBLE)
-typedef complex npy_cdouble;
+#ifdef __cplusplus
+extern "C++" {
+#include <complex>
+}
 #else
-typedef struct { double real, imag; } npy_cdouble;
+#include <complex.h>
 #endif
 
-#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_FLOAT)
-typedef complex float npy_cfloat;
-#else
-typedef struct { float real, imag; } npy_cfloat;
-#endif
-
-#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_LONG_DOUBLE)
-typedef complex long double npy_clongdouble;
-#else
-typedef struct {npy_longdouble real, imag;} npy_clongdouble;
-#endif
-#endif
 #if NPY_SIZEOF_COMPLEX_DOUBLE != 2 * NPY_SIZEOF_DOUBLE
 #error npy_cdouble definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { double real, imag; } npy_cdouble;
-
 #if NPY_SIZEOF_COMPLEX_FLOAT != 2 * NPY_SIZEOF_FLOAT
 #error npy_cfloat definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { float real, imag; } npy_cfloat;
-
 #if NPY_SIZEOF_COMPLEX_LONGDOUBLE != 2 * NPY_SIZEOF_LONGDOUBLE
 #error npy_clongdouble definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { npy_longdouble real, imag; } npy_clongdouble;
+
```

Comment:
Can any of the added things below go into a different header? I think some of the macros are more specific than `typedefs`. Same thing with the `PyArray_CLT` and friends in `numpy/core/include/numpy/ndarraytypes.h`

Reply:
Sure, agreed that it's not a good idea to keep these here. Which one do you think would be the most appropriate one? Maybe create a new header?

Reply:
If they need to be public, they could go into `npy_math.h`. If they can be internal-only, then maybe a npy_complex.h header in numpy/core/src/npymath?